internal/controller/templatemanagement_controller.go

```diff
@@ -48,7 +63,306 @@ func (r *TemplateManagementReconciler) Reconcile(ctx context.Context, req ctrl.R
+	}
+	if expectedState.serviceTemplatesState == nil {
+		expectedState.serviceTemplatesState = make(map[string]map[string]bool)
+	}
+	for _, rule := range rules {
+		var clusterTemplates []string
+		var serviceTemplates []string
+		namespaces, err := r.getNamespaces(ctx, rule.TargetNamespaces)
+		if err != nil {
+			return state{}, err
+		}
+		for _, ctChainName := range rule.ClusterTemplateChains {
+			ctChain := &hmc.ClusterTemplateChain{}
+			err := r.Get(ctx, types.NamespacedName{
+				Name: ctChainName,
+			}, ctChain)
+			if err != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+			for _, supportedTemplate := range ctChain.Spec.SupportedTemplates {
+				clusterTemplates = append(clusterTemplates, supportedTemplate.Name)
+			}
+		}
+		for _, stChainName := range rule.ServiceTemplateChains {
+			stChain := &hmc.ServiceTemplateChain{}
+			err := r.Get(ctx, types.NamespacedName{
+				Name: stChainName,
+			}, stChain)
+			if err != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+			for _, supportedTemplate := range stChain.Spec.SupportedTemplates {
+				serviceTemplates = append(serviceTemplates, supportedTemplate.Name)
+			}
+		}
+		for _, ns := range namespaces {
+			if expectedState.clusterTemplatesState[ns] == nil {
+				expectedState.clusterTemplatesState[ns] = make(map[string]bool)
+			}
+			if expectedState.serviceTemplatesState[ns] == nil {
+				expectedState.serviceTemplatesState[ns] = make(map[string]bool)
+			}
+			for _, ct := range clusterTemplates {
+				expectedState.clusterTemplatesState[ns][ct] = true
+			}
+			for _, st := range serviceTemplates {
+				expectedState.serviceTemplatesState[ns][st] = true
+			}
+		}
+	}
+	if errs != nil {
+		return state{}, errs
+	}
+	return expectedState, nil
+}
+
+func (r *TemplateManagementReconciler) getNamespaces(ctx context.Context, targetNamespaces hmc.TargetNamespaces) ([]string, error) {
+	if len(targetNamespaces.List) > 0 {
+		return targetNamespaces.List, nil
+	}
+	var selector labels.Selector
+	var err error
+	if targetNamespaces.StringSelector != "" {
+		selector, err = labels.Parse(targetNamespaces.StringSelector)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		selector, err = metav1.LabelSelectorAsSelector(targetNamespaces.Selector)
+		if err != nil {
+			return nil, fmt.Errorf("failed to construct selector from the namespaces selector %s: %w", targetNamespaces.Selector, err)
+		}
+	}
+	var listOpts *client.ListOptions
+	if selector != nil {
+		listOpts = &client.ListOptions{LabelSelector: selector}
+	}
+	namespaces := &v1.NamespaceList{}
+	err = r.List(ctx, namespaces, listOpts)
+	if err != nil {
+		return []string{}, err
+	}
+	result := make([]string, len(namespaces.Items))
+	for i, ns := range namespaces.Items {
+		result[i] = ns.Name
+	}
+	return result, nil
+}
+
+func (r *TemplateManagementReconciler) createTemplates(ctx context.Context, state map[string]map[string]bool, kind string) error {
```

Comment:
we might need to change the state structure here. if we need to distribute one template across x namespaces, we have to make x copyTemplate calls. I think it makes sense to inverse it and keep Templates as keys and two lists/a map of namespaces as values.